### PR TITLE
RGW: Getting an RGW service Segfault when assigning an attribute to an IAM role.

### PR DESCRIPTION
**Pull request by shreyanshjain7174**
The current implementation of RGWTagRole, which inherits RGWRestRole::verify_permission() from its base class, encounters a critical issue when loading RGWRole from storage and initializing the RGWRestRole::_role member variable.To address this issue and ensure that errors in initialization are appropriately handled, it is proposed to separate the initialization logic from the permission-checking logic.

Fixes: https://tracker.ceph.com/issues/64232
Signed-off-by: Shreyansh Sancheti <ssanchet@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [x] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/rgw/rgw_rest_role.cc [hunk 1]
```diff
@@ -112,6 +112,16 @@ int RGWRestRole::verify_permission(optional_yield y)
   return 0;
 }
 
+int RGWRestRole::init_processing(optional_yield y)
+{
+  op_ret = verify_permission(y);
```
Comment (cbodley) [inline src/rgw/rgw_rest_role.cc:117]: don't call `verify_permission()` directly here, just move the code that initializes `_role`:
```c++
  string role_name = s->info.args.get("RoleName");
  std::unique_ptr<rgw::sal::RGWRole> role = driver->get_role(role_name,
							    s->user->get_tenant());
  if (int ret = role->get(s, y); ret < 0) {
    if (ret == -ENOENT) {
      return -ERR_NO_ROLE_FOUND;
    }
    return ret;
  }
  _role = std::move(role);
  return 0;
```

**Review by cbodley — COMMENTED** [review]
don't forget this part:

> note that some ops (RGWCreateRole, RGWGetRole, RGWListRoles) provide their own verify_permission() to avoid calling it on the base class - so those ops should also avoid calling the base RGWRestRole::init_processing(). for example, with an override like:
```c++
int RGWCreateRole::init_processing(optional_yield y)
{
  return 0; // don't call RGWRestRole::init_processing()
}
```

**Comment by shreyanshjain7174** [comment]
@cbodley Do we need a unit test for this ?

File: src/rgw/rgw_rest_role.cc (hunk 1 above)
Comment (shreyanshjain7174) [reply src/rgw/rgw_rest_role.cc]:
```
int RGWRestRole::init_processing(optional_yield y)
{
    string role_name = s->info.args.get("RoleName");
    std::unique_ptr<rgw::sal::RGWRole> role = driver->get_role(role_name,
                                                               s->user->get_tenant());
    if (int ret = role->get(s, y); ret < 0)
    {
        if (ret == -ENOENT)
        {
            return -ERR_NO_ROLE_FOUND;
        }
        return ret;
    }
    _role = std::move(role);
    return 0;
}
```
added the method to initialise role inside the `init_processing`

**Comment by cbodley** [comment]
the ceph project has some guidelines for coding style in https://github.com/ceph/ceph/blob/main/CodingStyle

please indent with 2 spaces instead of 4

and please put opening braces `{` on the same line instead of a new line:
```c++
if (int ret = role->get(s, y); ret < 0) {
  if (ret == -ENOENT) {
```

**Comment by shreyanshjain7174** [comment]
- `jenkins test windows`

File: src/rgw/rgw_rest_role.cc [hunk 2]
```diff
@@ -107,8 +97,36 @@ int RGWRestRole::verify_permission(optional_yield y)
...
+  return 0;
+}
+
+int RGWGetRole::init_processing(optional_yield y)
+{
+  return 0;
+}
+
+int RGWListRoles::init_processing(optional_yield y)
+{
```
Comment (cbodley) [inline src/rgw/rgw_rest_role.cc:129]: looks good, but please group these functions together with their classes

and it's worth adding a comment like
```c++
  return 0; // avoid calling RGWRestRole::init_processing()
```

Comment (shreyanshjain7174) [reply src/rgw/rgw_rest_role.cc]: Done!

**Review by cbodley — APPROVED** [review]

**Comment by cbodley** [comment]
passed qa in https://pulpito.ceph.com/cbodley-2024-02-11_16:11:11-rgw-wip-cbodley2-testing-distro-default-smithi/

**Pull request merged**